vmdb/lib/extensions/as_const_missing.rb

```diff
@@ -0,0 +1,74 @@
+module ActiveSupport
+  module Dependencies
+    module ModuleConstMissing
+      # Rails +const_missing+ only loads the requested constant and any of its
+      # parents. However, when using STI and having class hierarchies that are
+      # several levels deep, ActiveRecord requires that all descendants of a
+      # model are loaded to generate the correct query. Additionally, the
+      # methods +subclasses+ and +descendants+ require that all descendants of
+      # a model are loaded in order to work as expected.
+      #
+      # Example:
+      #   Given a class hierarchy of:
+      #
+      #     class Aaa < ActiveRecord::Base; end
+      #     class Bbb < Aaa; end
+      #     class Ccc < Bbb; end
+      #
+      #   Without this change, the following queries will be generated:
+      #
+      #     Aaa.count
+      #     # SELECT COUNT(*) FROM "aaas"
+      #     Bbb.count
+      #     # SELECT COUNT(*) FROM "aaas" WHERE "aaas"."type" IN ('Bbb')
+      #     Ccc.count
+      #     # SELECT COUNT(*) FROM "aaas" WHERE "aaas"."type" IN ('Ccc')
+      #
+      #   The 'Bbb' query is incorrect since +const_missing+ does not load the
+      #   child objects.
+      #   With this change, the following queries will be generated:
+      #
+      #     Aaa.count
+      #     # SELECT COUNT(*) FROM "aaas"
+      #     Bbb.count
+      #     # SELECT COUNT(*) FROM "aaas" WHERE "aaas"."type" IN ('Bbb', 'Ccc')
+      #     Ccc.count
+      #     # SELECT COUNT(*) FROM "aaas" WHERE "aaas"."type" IN ('Ccc')
+      #
+      def const_missing_with_sti(constant)
+        const_missing_without_sti(constant).tap do
+          class_inheritance_relationships[constant.to_s].each { |c| load_subclass(c) }
+        end
+      end
+      alias_method_chain :const_missing, :sti
+
+      private
+
+      ClassInheritanceRelationship = Struct.new(:subclass, :file_path)
+
+      def self.class_inheritance_relationships
+        @class_inheritance_relationships ||= begin
+          patterns = Dependencies.autoload_paths.collect { |p| File.join(p, "*.rb") }
+          relats   = Hash.new { |h, k| h[k] = [] }
+          Dir.glob(patterns).each_with_object(relats) do |file, h|
+            File.read(file).lines.each do |l|
```

Comment:
We're open to suggestions...couldn't think of anything to avoid that.

Additionally, this doesn't handle namespacing, but right now none of our models do that.  I know @matthewd's PR will do that eventually, so we'll have to solve the problem, but maybe we can wait until then.